File: internal/api/handler.go
package api

import (
	"fmt"
	"net/http"

	"github.com/emicklei/go-restful/v3"
	"github.com/rs/zerolog"

	"github.com/CrowdStrike/codestrike/internal/config"
	"github.com/CrowdStrike/codestrike/internal/llm"
	"github.com/CrowdStrike/codestrike/internal/review"
	"github.com/CrowdStrike/codestrike/internal/setup"
	"github.com/CrowdStrike/codestrike/internal/tokenizer"
)

// Handler holds shared dependencies for all API endpoints.
type Handler struct {
	llmClient llm.LLMClient
	appConfig *config.Config
	envConfig *setup.Config
	logger    *zerolog.Logger
}

// NewHandler creates a new API handler with shared dependencies.
func NewHandler(llmClient llm.LLMClient, appConfig *config.Config, envConfig *setup.Config, log *zerolog.Logger) *Handler {
	return &Handler{
		llmClient: llmClient,
		appConfig: appConfig,
		envConfig: envConfig,
		logger:    log,
	}
}

// Healthz handles GET /api/v1/healthz.
func (h *Handler) Healthz(req *restful.Request, resp *restful.Response) {
	healthResponse := HealthResponse{
		Status:  "ok",
		Version: "1.0.0",
	}

	_ = resp.WriteHeaderAndEntity(http.StatusOK, healthResponse)
}

// Review handles POST /api/v1/review.
func (h *Handler) Review(req *restful.Request, resp *restful.Response) {
	var body ReviewRequest
	if err := req.ReadEntity(&body); err != nil {
		_ = resp.WriteHeaderAndEntity(http.StatusBadRequest, ErrorResponse{
			Error: fmt.Sprintf("invalid request body: %v", err),
		})
		return
	}

	if body.PrUrl == "" {
		_ = resp.WriteHeaderAndEntity(http.StatusBadRequest, ErrorResponse{
			Error: "pr_url is required",
		})
		return
	}

	ref, err := review.ParsePrUrl(body.PrUrl)
	if err != nil {
		_ = resp.WriteHeaderAndEntity(http.StatusBadRequest, ErrorResponse{
			Error: fmt.Sprintf("invalid PR URL: %v", err),
		})
		return
	}

	scmClient, err := setup.CreateSCMClient(h.envConfig, h.appConfig, ref)
	if err != nil {
		_ = resp.WriteHeaderAndEntity(http.StatusBadRequest, ErrorResponse{
			Error: fmt.Sprintf("creating SCM client: %v", err),
		})
		return
	}

	tok := tokenizer.NewForModel(h.appConfig.Review.Context.TokenizerModel)
	pipeline := review.NewPipeline(scmClient, h.llmClient, h.appConfig, tok, h.logger, review.Options{
		FullContext: body.FullContext,
	})

	if err := pipeline.Run(req.Request.Context(), ref); err != nil {
		h.logger.Error().Err(err).Str("pr_url", body.PrUrl).Msg("review failed")
		_ = resp.WriteHeaderAndEntity(http.StatusInternalServerError, ErrorResponse{
			Error: fmt.Sprintf("review failed: %v", err),
		})
		return
	}

	_ = resp.WriteHeaderAndEntity(http.StatusOK, ReviewResponse{
		Status: "reviewed",
		Owner:  ref.Owner,
		Repo:   ref.Repo,
		PR:     ref.Number,
	})
}

File: src/api/handler.ts
import type { Logger } from '../logger.js';
import type { Config } from '../config/config.js';
import type { LLMClient } from '../llm/types.js';
import type { EnvConfig } from '../setup.js';
import { createSCMClient } from '../setup.js';
import { parsePrUrl } from '../review/parser.js';
import { Pipeline } from '../review/pipeline.js';
import { newForModel } from '../tokenizer.js';
import type { HealthResponse, ReviewRequest, ReviewResponse, ErrorResponse } from './types.js';

export class Handler {
  private llmClient: LLMClient;
  private appConfig: Config;
  private envConfig: EnvConfig;
  private logger: Logger;

  constructor(llmClient: LLMClient, appConfig: Config, envConfig: EnvConfig, logger: Logger) {
    this.llmClient = llmClient;
    this.appConfig = appConfig;
    this.envConfig = envConfig;
    this.logger = logger;
  }

  healthz(): HealthResponse {
    return { status: 'ok', version: '1.0.0' };
  }

  async review(body: ReviewRequest): Promise<{ status: number; body: ReviewResponse | ErrorResponse }> {
    if (!body.pr_url) {
      return { status: 400, body: { error: 'pr_url is required' } };
    }

    let ref;
    try {
      ref = parsePrUrl(body.pr_url);
    } catch (err) {
      return { status: 400, body: { error: `invalid PR URL: ${(err as Error).message}` } };
    }

    let scmClient;
    try {
      scmClient = createSCMClient(this.envConfig, this.appConfig, ref);
    } catch (err) {
      return { status: 400, body: { error: `creating SCM client: ${(err as Error).message}` } };
    }

    const tok = newForModel(this.appConfig.review.context.tokenizer_model);
    const pipeline = new Pipeline(scmClient, this.llmClient, this.appConfig, tok, this.logger, {
      fullContext: body.full_context ?? false,
      dryRun: false,
    });

    try {
      await pipeline.run(ref);
    } catch (err) {
      this.logger.error({ err, pr_url: body.pr_url }, 'review failed');
      return { status: 500, body: { error: `review failed: ${(err as Error).message}` } };
    }

    return {
      status: 200,
      body: { status: 'reviewed', owner: ref.owner, repo: ref.repo, pr: ref.number },
    };
  }
}
